src/bun.js/test/snapshot.zig

```diff
@@ -0,0 +1,292 @@
+    update_snapshot: bool = false,
+
+    pub var update_all_snapshots: bool = false;
+
+    pub const TestData = struct {
+        k: ZigString,
+        v: JSC.JSValue,
+        count: u32,
+
+        pub const List = std.ArrayListUnmanaged(TestData);
+
+        fn lessThan(_: void, a: TestData, b: TestData) bool {
+            return ZigString.cmpDesc(void{}, a.k, b.k);
+        }
+    };
+
+    pub const SnapshotError = error{ SnapshotNotFound, CaseNotFound };
+
+    pub const CountersMap = std.HashMap(u64, u32, IdentityContext(u64), 80);
+
+    pub fn readAndParseSnapshot(this: *SnapshotFile, globalObject: *JSC.JSGlobalObject) void {
+        var snapshot_contents = this.readSnapshot() catch return;
+        const contents_string = ZigString.init(snapshot_contents);
+        this.parseSnapshot(&contents_string, globalObject) catch unreachable;
+    }
+
+    pub fn readSnapshot(this: *SnapshotFile) !bun.string {
+        var file: std.fs.File = std.fs.cwd().openFile(this.path.text, .{ .mode = .read_write }) catch {
+            this.snapshot_data = SnapshotError.SnapshotNotFound;
+            return SnapshotError.SnapshotNotFound;
+        };
+        const file_size = try file.getEndPos();
+        var snapshot_contents = try this.allocator.alloc(u8, file_size);
+        _ = try file.read(snapshot_contents);
+        file.close();
+        return snapshot_contents;
+    }
+
+    pub fn setSnapshotContents(this: *SnapshotFile, snapshot: js.JSValueRef) void {
+        this.snapshot_data = snapshot;
+    }
+
+    // Parses an existing by passing the snapshot file's contents into the javascript equivalent code:
+    //      f = new Function("exports", snapshotContents);
+    //      const param = {};
+    //      f(param);
+    // if the test snapshot file looks like: exports[`test name 1`] = `5`
+    // the value of param will be set to: { test name 1: '5' }
+    pub fn parseSnapshot(this: *SnapshotFile, snapshotContents: *const ZigString, globalObject: *JSC.JSGlobalObject) !void {
+        // Arguments to create function
+        var function_name = JSC.C.JSStringCreateStatic("anonymous", "anonymous".len);
+        var param_name = JSC.C.JSStringCreateStatic("exports", "exports".len);
+        var functionBody = JSC.C.JSStringCreateStatic(snapshotContents.ptr, snapshotContents.len);
+        var func_url = JSC.C.JSStringCreateStatic("file:///snapshot.zig", "file:///snapshot.zig".len);
+        var exception_ptr: ?[*]JSC.JSValueRef = null;
+
+        const function: js.JSObjectRef = JSC.C.JSObjectMakeFunction(globalObject, function_name, 1, &[_]JSC.C.JSStringRef{param_name}, functionBody, func_url, 1, exception_ptr);
+
+        var expect_arg: js.JSValueRef = JSC.JSValue.createEmptyObject(globalObject, 0).asRef();
+        const arguments = [1]js.JSValueRef{expect_arg};
+
+        // Call the function
+        _ = JSC.C.JSObjectCallAsFunction(globalObject, function, null, 1, &arguments, exception_ptr) orelse unreachable;
```

Comment:
The function `parseSnapshot` uses the [same logic as jest](https://github.com/facebook/jest/blob/9af2540f87e8e0d4f1e0528981529ad1bc30a28a/packages/jest-snapshot/src/utils.ts#L98-L130) to parse the snapshot file. 

Currently if the snapshot file is mis-formatted this will cause a segfault in bun. How can I add error handling so the test can fail normally if the snapshot is mis-formatted?

Reply:
Overview of how this function works:

If a snapshot file is formatted like:
```
exports[`test that the string is five`] = `5`;
```

The jest library / this snapshot logic will run the equivalent javascript code

```
      snapshotContents = fs.readFileSync(snapshotPath, 'utf8');
      // eslint-disable-next-line no-new-func
      const populate = new Function('exports', snapshotContents);
      populate(data);
```

Which will transform the data into an object that looks like the following:

```
exports = { `test that the string is five`: `5` }
```

However when the input snapshot file is incorrectly formatted it can cause the bun to segfault. For example if the snapshot file saved looks like:

```
`test that the string is five` = `5`;
```

Then the program will segfault at `snapshot.zig:149`. How can catch an exception from the zig call to `JSObjectCallAsFunction`?